sdk/src/secp256k1_instruction.rs

```diff
@@ -1,3 +1,790 @@
+//! be signed by a known secret key. For other uses cases, including many
+//! Ethereum-integration use cases, construction of the secp256k1 instruction
+//! must be done manually.
+//!
+//! # How to use this program
+//!
+//! Transactions that uses the secp256k1 native program will typically include
+//! at least two instructions: one for the secp256k1 program to verify the
+//! signatures, and one for a custom program that will check that the secp256k1
+//! instruction data matches what the program expects (using
+//! [`load_instruction_at_checked`] or [`get_instruction_relative`]). The
+//! signatures, messages, and Ethereum addresses being verified may reside in the
+//! instruction data of either of these instructions, or in the instruction data
+//! of one or more additional instructions, as long as those instructions are in
+//! the same transaction.
+//!
+//! [`load_instruction_at_checked`]: crate::sysvar::instructions::load_instruction_at_checked
+//! [`get_instruction_relative`]: crate::sysvar::instructions::get_instruction_relative
+//!
+//! Correct use of this program involves multiple steps, in client code and
+//! program code:
+//!
+//! - In the client:
+//!   - Sign the [`keccak`]-hashed messages with a secp256k1 ECDSA library,
+//!     like the [`libsecp256k1`] crate.
+//!   - Build any custom instruction data that contain signature, message, or
+//!     Ethereum address data that will be used by the secp256k1 instruction.
+//!   - Build the secp256k1 program instruction data, specifying the number of
+//!     signatures to verify, the instruction indexes within the transaction,
+//!     and offsets within those instruction's data, where the signatures,
+//!     messages, and Ethereum addresses are located.
+//!   - Build the custom instruction for the program that will check the results
+//!     of the secp256k1 native program.
+//!   - Package all instructions into a single transaction and submit them.
+//! - In the program:
+//!   - Load the secp256k1 instruction data with
+//!     [`load_instruction_at_checked`]. or [`get_instruction_relative`].
+//!   - Check that the secp256k1 program ID is equal to
+//!     [`secp256k1_program::ID`], so that the signature verification cannot be
+//!     faked with a malicious program.
+//!   - Check that the public keys and messages are the expected values per
+//!     the program's requirements.
+//!
+//! [`secp256k1_program::ID`]: crate::secp256k1_program::ID
+//!
+//! The signature, message, or Ethereum addresses may reside in the secp256k1
+//! instruction data itself as additional data, their bytes following the bytes
+//! of the protocol required by the secp256k1 instruction to locate the
+//! signature, message, and Ethereum address data. This is the technique used by
+//! `new_secp256k1_instruction` for simple signature verification.
+//!
+//! The `solana_sdk` crate provides few APIs for building the instructions and
```

Comment:
provides a

Reply:
I can change this, and it will still make sense, though it is written as intended, conveying that for many use cases the APIs do not exist.

Reply:
Yes, you are right.